# 4.8 Grouping Data & Aggregating Variables

## This script includes the following points:

#### 1. Import the data sets.
#### 2. Find the aggregated mean of the “order_number” column grouped by “department_id”.
#### 3. Analyze the difference between the results for the entire dataframe and those of the subset.
#### 4. Create a loyalty flag using the transform() and loc() functions.
#### 5. Analyze spending habits by loyalty category.
#### 6. Create a spending flag for each user based on the average price across all their orders.
#### 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior.
#### 8. Export the data as a pickle file.

# Importing Libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets.

In [9]:
path = r'/Users/test/Desktop/Data Analysis/11-2024 Instacart Basket Analysis'

In [11]:
# Import orders_products_merged_derived.pkl file from Exercise 4.7
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'orders_products_merged_derived.pkl'))

In [15]:
# Check dimensions of imported dataframe for consistency
df_ords_prods_merge.shape

(32399732, 19)

### 2. Find the aggregated mean of the “order_number” column grouped by “department_id”.

In [19]:
# Group by 'department_id' and Calculate Mean of 'order_number' using agg() Function
df_ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.667853
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 3. Analyze the difference between the results for the entire dataframe and those of the subset.

1). Compared to the results from the subset, the results from the entire dataframe includes all 'department_id' values.

2). Notable Differences: Some departments show significant changes in their mean values. For instance, Department 14 has a higher mean in the subset (19.24) than in the full dataframe (16.77), while Department 17 exhibits the opposite trend, increasing from 11.29 in the subset to 15.69 in the full dataset.

3). Consistent Departments: Some departments, such as 4, 7, 13 and 20, display minimal differences between the two datasets, indicating stable patterns. 

4). Overall, the entire dataframe highlights more diverse and frequent ordering behaviors across departments, providing insights that are not fully captured by the subset.

### 4. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [27]:
# Aggregate Data with transform() Function
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/23/dwh_pxj57x19sw0ns77c_swh0000gn/T/ipykernel_37651/3473307984.py:2: FutureWarning: The provided callable <function max at 0x1052965c0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [29]:
# Check the Output
df_ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    days_since_prior_order  first_order  add_to_cart_order  reordered _merge  \
0                      3.0        False                  5          0   both   
1                     20.0        False                  1          1   both   
2                      6.0        False                 20          0   both   
3                      NaN         True                 10          0   both   
4                      9.0        False                 11          1   both   
5                     26.0        False                  7          0   both   
6                     30.0        False                  2          0   both   
7                     30.0        False                  1       

In [31]:
# Create a loyalty flag with loc() function.
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [33]:
# Check value counts on 'loyalty_flag'.
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15874128
Loyal customer      10282763
New customer         6242841
Name: count, dtype: int64

In [35]:
# Check the updated output
df_ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   
5           1  Chocolate Sandwich Cookies        61             19     5.8   
6           1  Chocolate Sandwich Cookies        61             19     5.8   
7           1  Chocolate Sandwich Cookies        61             19     5.8   
8           1  Chocolate Sandwich Cookies        61             19     5.8   
9           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   ...  first_order  add_to_cart_order  reordered  _merge    price_range_loc  \
0  ...        False                  5          0    both  Mid-range product   
1  ...        False                  1          1    both  Mid-range product   
2  ...        False                 20          0    both  Mid-range product   
3  ...         True                 10          0    both  Mid-range product   
4  ...        False                 11          1    both  Mid-range product   
5  ...        False                  7          0    both  Mid-range product   
6  ...        False                  2          0    both  Mid-range product   
7  ...        False                  1          0    both  Mid-range product   
8  ...        False                  1          0    both  Mid-range product   
9  ...        False                  2          1    both  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy           Most orders        32   
1  Regularly busy  Regularly busy        Average orders        32   
2     Busiest day    Busiest days        Average orders         5   
3  Regularly busy    Slowest days           Most orders         3   
4      Least busy    Slowest days        Average orders         3   
5  Regularly busy    Busiest days        Average orders        26   
6  Regularly busy  Regularly busy           Most orders         9   
7  Regularly busy    Slowest days           Most orders        12   
8  Regularly busy    Busiest days        Average orders        20   
9  Regularly busy    Busiest days           Most orders        20   

       loyalty_flag  
0  Regular customer  
1  Regular customer  
2      New customer  
3      New customer  
4      New customer  
5  Regular customer  
6      New customer  
7  Regular customer  
8  Regular customer  
9  Regular customer  

[10 rows x 21 columns]

### 5. Analyze spending habits by loyalty category.

In [38]:
# Aggregate the data based on prices, providing basic statistics (mean, min and max).
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

While there is a slight difference in the average price of products purchased by different customer groups, the minimum and maximum prices remain consistent. This suggests that purchasing behavior in terms of price does not vary significantly between loyal, regular, and new customers.

### 6. Create a spending flag for each user based on the average price across all their orders.

1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [44]:
# Create a new column for the average spending of a customer
df_ords_prods_merge['average_spend'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/23/dwh_pxj57x19sw0ns77c_swh0000gn/T/ipykernel_37651/2089113553.py:2: FutureWarning: The provided callable <function mean at 0x105296fc0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['average_spend'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [46]:
# Check the new column
df_ords_prods_merge[['user_id', 'average_spend']].head(20)

user_id  average_spend
0       138       6.935811
1       138       6.935811
2       709       7.930208
3       764       4.972414
4       764       4.972414
5       777       6.935398
6       825       5.957576
7       910       6.680000
8      1052       7.162500
9      1052       7.162500
10     1379       5.813636
11     1480       5.081633
12     1480       5.081633
13     1480       5.081633
14     1494       6.803797
15     1494       6.803797
16     1494       6.803797
17     1540       6.607122
18     1540       6.607122
19     1540       6.607122

In [48]:
# Create a 'spending flag' with loc() Function.
df_ords_prods_merge.loc[df_ords_prods_merge['average_spend'] >= 10, 'spending_flag'] = 'High spender'
df_ords_prods_merge.loc[df_ords_prods_merge['average_spend'] < 10, 'spending_flag'] = 'Low spender'

In [50]:
# Check spender flags
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32280013
High spender      119719
Name: count, dtype: int64

The results indicate that the number of high spenders is relatively small.

### 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior.

The criteria for the flag should be as follows:
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a 'Non-frequent customer'
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a 'Regular customer'
3. If the median is lower than or equal to 10, then the customer should be labeled a 'Frequent customer'

In [56]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column.
df_ords_prods_merge['customer_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [58]:
# Check the relevant columns 
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'customer_frequency']].head(20)

user_id  days_since_prior_order  customer_frequency
0       138                     3.0                 8.0
1       138                    20.0                 8.0
2       709                     6.0                 8.0
3       764                     NaN                 9.0
4       764                     9.0                 9.0
5       777                    26.0                11.0
6       825                    30.0                20.0
7       910                    30.0                 6.0
8      1052                    19.0                10.0
9      1052                    15.0                10.0
10     1379                     NaN                20.0
11     1480                     7.0                 7.0
12     1480                     7.0                 7.0
13     1480                    12.0                 7.0
14     1494                     2.0                 7.0
15     1494                     6.0                 7.0
16     1494                    10.0                 7.0
17     1540                     7.0                 7.0
18     1540                     6.0                 7.0
19     1540                    14.0                 7.0

In [60]:
# Create a 'order frequency flag' with loc() function.
df_ords_prods_merge.loc[df_ords_prods_merge['customer_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['customer_frequency'] <= 20) & (df_ords_prods_merge['customer_frequency'] > 10), 'order_frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['customer_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [62]:
# Check the value counts for new column. 
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21556644
Regular customer          7207340
Non-frequent customer     3635743
NaN                             5
Name: count, dtype: int64

In [64]:
# Check relevant columns of updated dataframe.
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'customer_frequency', 'order_frequency_flag']].head(20)

user_id  days_since_prior_order  customer_frequency order_frequency_flag
0       138                     3.0                 8.0    Frequent customer
1       138                    20.0                 8.0    Frequent customer
2       709                     6.0                 8.0    Frequent customer
3       764                     NaN                 9.0    Frequent customer
4       764                     9.0                 9.0    Frequent customer
5       777                    26.0                11.0     Regular customer
6       825                    30.0                20.0     Regular customer
7       910                    30.0                 6.0    Frequent customer
8      1052                    19.0                10.0    Frequent customer
9      1052                    15.0                10.0    Frequent customer
10     1379                     NaN                20.0     Regular customer
11     1480                     7.0                 7.0    Frequent customer
12     1480                     7.0                 7.0    Frequent customer
13     1480                    12.0                 7.0    Frequent customer
14     1494                     2.0                 7.0    Frequent customer
15     1494                     6.0                 7.0    Frequent customer
16     1494                    10.0                 7.0    Frequent customer
17     1540                     7.0                 7.0    Frequent customer
18     1540                     6.0                 7.0    Frequent customer
19     1540                    14.0                 7.0    Frequent customer

### 8. Export the data as a pickle file.

In [67]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_grouped.pkl'))